# Softwares used in the annotation module

## [add_mosdepth_coverage_to_gvcf](https://github.com/hydra-genetics/annotation/blob/develop/workflow/scripts/add_mosdepth_coverage_to_gvcf.py)
In house script that adds coverage per position based on coverage data from mosdepth to a genome vcf file created by Mutect2. Data is added to the FORMAT field. 

### :snake: Rule

#SNAKEMAKE_RULE_SOURCE__add_mosdepth_coverage_to_gvcf__add_mosdepth_coverage_to_gvcf#

#### :left_right_arrow: input / output files

#SNAKEMAKE_RULE_TABLE__add_mosdepth_coverage_to_gvcf__add_mosdepth_coverage_to_gvcf#

### :wrench: Configuration

#### Software settings (`config.yaml`)

#CONFIGSCHEMA__add_mosdepth_coverage_to_gvcf#

#### Resources settings (`resources.yaml`)

#RESOURCESSCHEMA__add_mosdepth_coverage_to_gvcf#

---

## [add_multi_snv_in_codon](https://github.com/hydra-genetics/annotation/blob/develop/workflow/scripts/add_multi_snv_in_codon.py)
In house script that goes through a vcf file and based on the VEP annotation finds multiple SNVs in the same codon. When these are found a merged variant are added to the vcf file. After this step the vcf file should be sorted and then reannotated with vep.

### :snake: Rule

#SNAKEMAKE_RULE_SOURCE__add_multi_snv_in_codon__add_multi_snv_in_codon#

#### :left_right_arrow: input / output files

#SNAKEMAKE_RULE_TABLE__add_multi_snv_in_codon__add_multi_snv_in_codon#

### :wrench: Configuration

#### Software settings (`config.yaml`)

#CONFIGSCHEMA__add_multi_snv_in_codon#

#### Resources settings (`resources.yaml`)

#RESOURCESSCHEMA__add_multi_snv_in_codon#

---

## [annotate_cnv](https://github.com/hydra-genetics/annotation/blob/develop/workflow/scripts/annotate_cnv.py)
In house script that annotates cnv variants with gene names based on a annotated bed file. 

### :snake: Rule

#SNAKEMAKE_RULE_SOURCE__annotate_cnv__annotate_cnv#

#### :left_right_arrow: input / output files

#SNAKEMAKE_RULE_TABLE__annotate_cnv__annotate_cnv#

### :wrench: Configuration

#### Software settings (`config.yaml`)

#CONFIGSCHEMA__annotate_cnv#

#### Resources settings (`resources.yaml`)

#RESOURCESSCHEMA__annotate_cnv#

---

## [artifact_annotation](https://github.com/hydra-genetics/annotation/blob/develop/workflow/scripts/artifact_annotation.py)
In house script that annotates small variants in a vcf file with artifact information based on a artifact file generated by the reference module usually from a pool of normals.

### :snake: Rule

#SNAKEMAKE_RULE_SOURCE__artifact_annotation__artifact_annotation#

#### :left_right_arrow: input / output files

#SNAKEMAKE_RULE_TABLE__artifact_annotation__artifact_annotation#

### :wrench: Configuration

#### Software settings (`config.yaml`)

#CONFIGSCHEMA__artifact_annotation#

#### Resources settings (`resources.yaml`)

#RESOURCESSCHEMA__artifact_annotation#

---

## [background_annotation](https://github.com/hydra-genetics/annotation/blob/develop/workflow/scripts/background_annotation.py)
In house script that annotates small variants in a vcf file with background noise information based on a background file generated by the reference module usually from a pool of normals.

### :snake: Rule

#SNAKEMAKE_RULE_SOURCE__background_annotation__background_annotation#

#### :left_right_arrow: input / output files

#SNAKEMAKE_RULE_TABLE__background_annotation__background_annotation#

### :wrench: Configuration

#### Software settings (`config.yaml`)

#CONFIGSCHEMA__background_annotation#

#### Resources settings (`resources.yaml`)

#RESOURCESSCHEMA__background_annotation#

---

## [bcftools_annotate](https://samtools.github.io/bcftools/bcftools.html#annotate)
Annotate a `vcf` file using bcftools annotate. The annotation is obtained from a supplied annotation database and the annotation added is described by the annotation string.

### :snake: Rule

#SNAKEMAKE_RULE_SOURCE__bcftools__bcftools_annotate#

#### :left_right_arrow: input / output files

#SNAKEMAKE_RULE_TABLE__bcftools__bcftools_annotate#

### :wrench: Configuration

#### Software settings (`config.yaml`)

#CONFIGSCHEMA__bcftools_annotate#

#### Resources settings (`resources.yaml`)

#RESOURCESSCHEMA__bcftools_annotate#

---

## [bcftools_sort](https://samtools.github.io/bcftools/bcftools.html#sort)
Sort a `vcf` file using bcftools sort. 

### :snake: Rule

#SNAKEMAKE_RULE_SOURCE__bcftools__bcftools_sort#

#### :left_right_arrow: input / output files

#SNAKEMAKE_RULE_TABLE__bcftools__bcftools_sort#

### :wrench: Configuration

#### Software settings (`config.yaml`)

#CONFIGSCHEMA__bcftools_sort#

#### Resources settings (`resources.yaml`)

#RESOURCESSCHEMA__bcftools_sort#

---

## [bgzip_vcf](http://www.htslib.org/doc/bgzip.html)
Compress a `.vcf`-file using bgzip.

### :snake: Rule

#SNAKEMAKE_RULE_SOURCE__bgzip__bgzip_vcf#

#### :left_right_arrow: input / output files

#SNAKEMAKE_RULE_TABLE__bgzip__bgzip_vcf#

### :wrench: Configuration

#### Software settings (`config.yaml`)

#CONFIGSCHEMA__bgzip_vcf#

#### Resources settings (`resources.yaml`)

#RESOURCESSCHEMA__bgzip_vcf#

---

## [calculate_seqrun_background](https://github.com/hydra-genetics/annotation/blob/develop/workflow/scripts/calculate_background.py)
Calculated the background noise level in a specific sequencing run based on genome vcf files generated by the rule [gatk_mutect2_gvcf](https://github.com/hydra-genetics/snv_indels/blob/develop/workflow/rules/gatk.smk).

### :snake: Rule

#SNAKEMAKE_RULE_SOURCE__calculate_seqrun_background__calculate_seqrun_background#

#### :left_right_arrow: input / output files

#SNAKEMAKE_RULE_TABLE__calculate_seqrun_background__calculate_seqrun_background#

### :wrench: Configuration

#### Software settings (`config.yaml`)

#CONFIGSCHEMA__calculate_seqrun_background#

#### Resources settings (`resources.yaml`)

#RESOURCESSCHEMA__calculate_seqrun_background#

---

## [hotspot_annotation](https://github.com/hydra-genetics/annotation/blob/develop/workflow/scripts/add_hotspot_information.py)
In house script that annotates small variants in a vcf file with hotspot information from a text file describing different hotspot variants of interest. 

### :snake: Rule

#SNAKEMAKE_RULE_SOURCE__hotspot_annotation__hotspot_annotation#

#### :left_right_arrow: input / output files

#SNAKEMAKE_RULE_TABLE__hotspot_annotation__hotspot_annotation#

### :wrench: Configuration

#### Software settings (`config.yaml`)

#CONFIGSCHEMA__hotspot_annotation#

#### Resources settings (`resources.yaml`)

#RESOURCESSCHEMA__hotspot_annotation#

---

## [simple_sv_annotation](https://github.com/AstraZeneca-NGS/simple_sv_annotation)
Simplifies annotations of vcf files annotated with SnpEff. Uses a gene list and known fusion partners.

### :snake: Rule

#SNAKEMAKE_RULE_SOURCE__simple_sv_annotation__simple_sv_annotation#

#### :left_right_arrow: input / output files

#SNAKEMAKE_RULE_TABLE__simple_sv_annotation__simple_sv_annotation#

### :wrench: Configuration

#### Software settings (`config.yaml`)

#CONFIGSCHEMA__simple_sv_annotation#

#### Resources settings (`resources.yaml`)

#RESOURCESSCHEMA__simple_sv_annotation#

---

## [snpeff](http://pcingola.github.io/SnpEff/)
Annotate vcf files using SnpEff.

### :snake: Rule

#SNAKEMAKE_RULE_SOURCE__snpeff__snpeff#

#### :left_right_arrow: input / output files

#SNAKEMAKE_RULE_TABLE__snpeff__snpeff#

### :wrench: Configuration

#### Software settings (`config.yaml`)

#CONFIGSCHEMA__snpeff#

#### Resources settings (`resources.yaml`)

#RESOURCESSCHEMA__snpeff#

---

## [stranger](https://github.com/Clinical-Genomics/stranger)
Annotates output files from ExpansionHunter with the pathologic implications of the repeat sizes.

### :snake: Rule

#SNAKEMAKE_RULE_SOURCE__stranger__stranger#

#### :left_right_arrow: input / output files

#SNAKEMAKE_RULE_TABLE__stranger__stranger#

### :wrench: Configuration

#### Software settings (`config.yaml`)

#CONFIGSCHEMA__stranger#

#### Resources settings (`resources.yaml`)

#RESOURCESSCHEMA__stranger#

---

## [tabix_vcf](http://www.htslib.org/doc/tabix.html)
Creates an index file for faster processing of positions in a bgzipped vcf file. Must be sorted.

### :snake: Rule

#SNAKEMAKE_RULE_SOURCE__bgzip__tabix_vcf#

#### :left_right_arrow: input / output files

#SNAKEMAKE_RULE_TABLE__bgzip__tabix_vcf#

### :wrench: Configuration

#### Software settings (`config.yaml`)

#CONFIGSCHEMA__tabix_vcf#

#### Resources settings (`resources.yaml`)

#RESOURCESSCHEMA__tabix_vcf#

---

## [vep](http://www.htslib.org/doc/samtools-merge.html)
Annotates vcf file using Ensemble Variant Effect Predictor (VEP). Uses local caches of databases to speed up annotation. Annotation options are added as string in the extra config parameter to VEP.

### :snake: Rule

#SNAKEMAKE_RULE_SOURCE__vep__vep#

#### :left_right_arrow: input / output files

#SNAKEMAKE_RULE_TABLE__vep__vep#

### :wrench: Configuration

#### Software settings (`config.yaml`)

#CONFIGSCHEMA__vep#

#### Resources settings (`resources.yaml`)

#RESOURCESSCHEMA__vep#

---

## [whatshap_phase](https://whatshap.readthedocs.io/en/latest/index.html)
WhatsHap is a software for phasing genomic variants and haplotagging reads. It is especially suitable for long reads, but works also well with short reads.

### :snake: Rule

#SNAKEMAKE_RULE_SOURCE__whatshap__whatshap_phase#

#### :left_right_arrow: input / output files

#SNAKEMAKE_RULE_TABLE__whatshap__whatshap_phase#

### :wrench: Configuration

#### Software settings (`config.yaml`)

#CONFIGSCHEMA__whatshap_phase#

#### Resources settings (`resources.yaml`)

#RESOURCESSCHEMA__whatshap_phase#

---

## [whatshap_haplotag](https://whatshap.readthedocs.io/en/latest/index.html)
WhatsHap is a software for phasing genomic variants and haplotagging reads. It is especially suitable for long reads, but works also well with short reads.

### :snake: Rule

#SNAKEMAKE_RULE_SOURCE__whatshap__whatshap_haplotag#

#### :left_right_arrow: input / output files

#SNAKEMAKE_RULE_TABLE__whatshap__whatshap_haplotag#

### :wrench: Configuration

#### Software settings (`config.yaml`)

#CONFIGSCHEMA__whatshap_haplotag#

#### Resources settings (`resources.yaml`)

#RESOURCESSCHEMA__whatshap_haplotag#